docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,239 @@
+  HTTP POST request.  The server uses a custom implementation of JSON-RPC that
+  is not fully compatible with the [JSON-RPC 2.0 spec][json-rpc], but handles
+  the common cases.
+
+- Websocket: Queries may be sent as JSON-RPC request messages via a websocket.
+  This transport uses more or less the same JSON-RPC plumbing as the HTTP POST
+  handler.
+
+  TODO: There appear to be some methods that are _only_ exported via websocket.
+  Figure out what is going on there and why it's that way. I think these may be
+  here for the same reason as the gRPC subset, viz., to support external
+  clients who could now talk directly to the application.
+
+- gRPC: A subset of queries may be issued in protocol buffer format to the gRPC
+  interface described above under (4). As noted, this endpoint is deprecated
+  and will be removed in v0.36.
+
+### Opportunities for Simplification
+
+**Claim:** There are too many IPC mechanisms.
+
+The preponderance of ABCI usage is via the Cosmos SDK, which means the
+application and the consensus node are compiled together into a single binary,
+and the consensus node calls the ABCI methods of the application directly as Go
+functions.
+
+TODO: What are the remaining use cases for the socket protocol and gRPC? Ethan
+mentioned some cases like web-based explorer UIs and possibly other tools.
+
+As far as I can tell the primary consumers of the multi-headed "RPC service"
+are the light client and the `tendermint` command-line client. There is
+probably some local use via curl, but I expect that is mostly ad hoc. Ethan
+reports that nodes are often configured with the ports to the RPC service
+blocked, which is good for security but complicates use by the light client.
+
+TODO: Figure out what is going on with the weird pseudo-inheritance service
+interface.  The BaseService type makes sense, but its embeds are more
+complicated.
+
+### Context: ABCI Issues
+
+In the original design of ABCI, the presumption was that all access to the
+application would be mediated by the consensus node. This makes sense, since
+outside access could change application state and corrupt the consensus
+process. Of course, even without outside access an application could behave
+nondeterministically, but allowing other programs to send it requests was seen
+as inviting trouble.
+
+Later, however, it was noted that most of the time, tools specific to a
+particular application don't really want to talk to the consensus module
+directly. The application is the "owner" of the state machine the consensus
+engine is replicating, so tools that want to know things about application
+state should talk to the application. Otherwise, such tools would have to bake
+in knowledge about Tendermint (e.g., its interfaces and data structures) that
+they might otherwise not need or care about.
+
+However, this raises another issue: The consensus node is the ABCI client, so
+it is inconvenient for the application to "push" work into the consensus
```

Comment:
I'm a bit unclear on what kind of work the application would need to push into Tendermint at this time. In my mental model, Tendermint is a deterministic system, so I'm not certain where it would be performing additional computation on behalf of the application, mind clarifying or adding some examples?

Reply:
I'm not entirely clear on it myself. This is distilled out of our discussion with @ebuchman the other day. The cases I know about now are for bootstrapping and debugging, and are currently handled by the RPC plumbing. But Ethan alluded to the possibility of others—I don't know if those are real, hypothetical, or in-development.

Reply:
I attempted to clarify this a bit. I still don't know what the use cases are, but the Cosmos SDK works around the directionality issue using the RPC service ("the way it does will shock you!") to poke transactions in. For our purposes, that isn't so bad, since the application and the consensus node should always be adjacent even if they're in separate processes. It's a little ugly, but not in a way that is deeply harmful to the design. And, if ABCI evolves to handle this case explicitly, we could rip it out later.